lib/ipam/ipam.go

```diff
@@ -1598,3 +1636,62 @@ func (c ipamClient) GetUtilization(ctx context.Context, args GetUtilizationArgs)
 	}
 	return usage, nil
 }
+
+// GetPoolDetails returns details about blocks.
+func (c ipamClient) GetPoolDetails(ctx context.Context, args GetUtilizationArgs) ([]*PoolDetails, error) {
+	var details []*PoolDetails
+
+	// Read all pools.
+	allPools, err := c.pools.GetAllPools()
+	if err != nil {
+		log.WithError(err).Errorf("Error getting IP pools")
+		return nil, err
+	}
+
+	// Identify the ones we want and create a PoolUtilization for each of those.
+	wantAllPools := len(args.Pools) == 0
+	wantedPools := set.FromArray(args.Pools)
+	for _, pool := range allPools {
+		if wantAllPools ||
+			wantedPools.Contains(pool.Name) ||
+			wantedPools.Contains(pool.Spec.CIDR) {
+			details = append(details, &PoolDetails{
+				Name: pool.Name,
+				CIDR: net.MustParseNetwork(pool.Spec.CIDR).IPNet,
+			})
+		}
+	}
+
+	// Read details for all blocks.
+	blocks, err := c.client.List(ctx, model.BlockListOptions{}, "")
+	if err != nil {
+		return nil, err
+	}
+	for _, kvp := range blocks.KVPairs {
+		b := kvp.Value.(*model.AllocationBlock)
+		log.Debugf("Got block: %v", b)
+
+		// Find which pool this block belongs to.
+		for _, poolDetails := range details {
+			if b.CIDR.IsNetOverlap(poolDetails.CIDR) {
+				log.Debugf("Block CIDR %v belongs to pool %v", b.CIDR, poolDetails.Name)
+
+				blockDetail := &BlockDetail{CIDR: b.CIDR.IPNet, Node: *b.Affinity}
```

Comment:
`b.Affinity` might be nil, so we shouldn't do a pointer dereference here without checking that.